### PR TITLE
[GPU] Improve memory usage management to distinguish allocation type

### PR DESCRIPTION
**Pull request by andrew-k-park**
### Details:
 - Currently it is accumulating and subtracting for a shared memory record
 - Update engine API to distinguish used memory by allocation type

### Tickets:
 - 63233

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp [hunk 1]
```diff
@@ -123,19 +123,36 @@ uint64_t engine::get_max_used_device_memory() const {
...
+void engine::add_memory_used(size_t bytes, allocation_type type) {
+    auto iter = memory_usage_map.find(type);
+    if (iter == memory_usage_map.end()) {
+        memory_usage_map.insert(std::make_pair(type, 0));
+    }
+    memory_usage_map[type] += bytes;
+
+    uint64_t memory_usage{};
+    for (auto const& m : memory_usage_map) {
+        memory_usage += m.second.load();
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/runtime/engine.cpp:144]: why is this accumulating all types? 
I think peak_memory should be also managed separately...

File: inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp [hunk 2]
```diff
@@ -142,7 +142,7 @@ class engine {
     const device::ptr _device;
     engine_configuration _configuration;
 
-    std::atomic<uint64_t> memory_usage = {0};
+    std::map<allocation_type, std::atomic<uint64_t>> memory_usage_map;
     std::atomic<uint64_t> peak_memory_usage = {0};
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp:146]: I think peak memory should be managed separately too

File: inference-engine/thirdparty/clDNN/runtime/memory.cpp [hunk 3]
```diff
@@ -20,26 +20,26 @@ namespace cldnn {
...
-        _engine->subtract_memory_used(_bytes_count);
+        _engine->subtract_memory_used(_bytes_count, _type);
     }
 
     GPU_DEBUG_GET_INSTANCE(debug_config);
     GPU_DEBUG_IF(debug_config->verbose >= 1) {
-        GPU_DEBUG_COUT << "Free " << _bytes_count << " bytes"
-                       << " (current=" << _engine->get_used_device_memory() << ";"
+        GPU_DEBUG_COUT << "Free " << _bytes_count << " bytes of " << _type << " allocation type"
+                       << " (current=" << _engine->get_used_device_memory(_type) << ";"
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/runtime/memory.cpp:42]: get_max_used_device_memory should be done for each alloc_type too

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp (hunk 1 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp]: done

File: inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp (hunk 2 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp]: done

File: inference-engine/thirdparty/clDNN/runtime/memory.cpp (hunk 3 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/memory.cpp]: done

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp [hunk 4]
```diff
@@ -119,23 +119,40 @@ memory_ptr engine::share_surface(const layout& layout, shared_surface surf, uint
...
     }
 }
 
-void engine::subtract_memory_used(size_t bytes) {
-    memory_usage -= bytes;
+void engine::subtract_memory_used(size_t bytes, allocation_type type) {
+    auto iter = memory_usage_map.find(type);
+    if (iter != memory_usage_map.end()) {
+        memory_usage_map[type] -= bytes;
+    }
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/runtime/engine.cpp:155]: Lets add an exception if the allocation type is not registered.

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp [hunk 5]
```diff
@@ -119,23 +119,40 @@ memory_ptr engine::share_surface(const layout& layout, shared_surface surf, uint
...
+    if (iter != peak_memory_usage_map.end()) {
+        peak_memory_usage = iter->second.load();
+    }
+    return peak_memory_usage;
 }
 
-uint64_t engine::get_used_device_memory() const {
-    return memory_usage.load();
+uint64_t engine::get_used_device_memory(allocation_type type) const {
+    uint64_t memory_usage{};
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/runtime/engine.cpp:132]: why not concretely set this as 0?

File: inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp [hunk 6]
```diff
@@ -96,17 +96,17 @@ class engine {
     /// Returns user context handle which was used to create the engine
     virtual void* get_user_context() const = 0;
 
-    /// Returns the maximum amount of GPU memory that engine allocated in current process
-    uint64_t get_max_used_device_memory() const;
+    /// Returns the maximum amount of GPU memory specified allocation @p type that engine allocated in current process
```
[the remark this answers is not in the document]
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/api/cldnn/runtime/engine.hpp]: done

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp (hunk 4 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp:157]: done

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp (hunk 5 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp]: Empty brace also means zero-initialization. but I've added it for now.

**Comment by yeonbok** [comment]
LGTM

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp [hunk 7]
```diff
@@ -119,23 +119,42 @@ memory_ptr engine::share_surface(const layout& layout, shared_surface surf, uint
...
 }
 
-void engine::subtract_memory_used(size_t bytes) {
-    memory_usage -= bytes;
+void engine::subtract_memory_used(size_t bytes, allocation_type type) {
+    auto iter = memory_usage_map.find(type);
+    if (iter != memory_usage_map.end()) {
+        memory_usage_map[type] -= bytes;
+    } else {
+        throw std::invalid_argument("Invalid allocation type");
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/runtime/engine.cpp:156]: More specifically, "Attempting free of unallocated memory"?

Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp]: done

File: inference-engine/thirdparty/clDNN/tests/test_cases/memory_test.cpp [hunk 8]
```diff
@@ -403,7 +403,7 @@ TEST(memory_pool, shared_mem_pool_diff_batches) {
     network network_second(*engine, topo, bo);
     network_second.set_input_data("input", input_1);
     auto outputs_second = network_second.execute();
-    EXPECT_EQ(engine->get_max_used_device_memory(), (uint64_t)3928);
+    EXPECT_EQ(engine->get_max_used_device_memory(), (uint64_t)4328);
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/tests/test_cases/memory_test.cpp:406]: Why? Do we skip some buffer in current version?

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp [hunk 9]
```diff
@@ -120,22 +120,49 @@ memory_ptr engine::share_surface(const layout& layout, shared_surface surf, uint
...
     }
+    return memory_usage;
 }
 
-void engine::subtract_memory_used(size_t bytes) {
-    memory_usage -= bytes;
+void engine::add_memory_used(size_t bytes, allocation_type type) {
+    if (!memory_usage_map.count(type) && !peak_memory_usage_map.count(type)) {
+        memory_usage_map.insert(std::make_pair(type, 0));
+        peak_memory_usage_map.insert(std::make_pair(type, 0));
```
Comment (vladimir-paramuzov) [inline inference-engine/thirdparty/clDNN/runtime/engine.cpp:151]: This part is not thread safe. Please check add/subtract impls to ensure thread safety of all operations.

File: inference-engine/thirdparty/clDNN/tests/test_cases/memory_test.cpp (hunk 8 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/tests/test_cases/memory_test.cpp:406]: Currently, the peak memory usage has been changed to be managed separately by allocation type, so ground truth value in this test case(iteratively executing the updated topology) needs to be changed.

File: inference-engine/thirdparty/clDNN/runtime/engine.cpp (hunk 9 above)
Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp:153]: done

Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp:153]: @andrew-k-park  Could you try multi-threaded application by yourself? e.g., fix benchmark app to run multiple threads inside it

Comment (andrew-k-park) [reply inference-engine/thirdparty/clDNN/runtime/engine.cpp]: As we discussed, I've confirmed thread safety when I try to run multi-threaded sample app, memory map works well as intended.

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**